File: ts/test/budget.test.ts
/** Offline tests for reading a compute budget across all three versions. */

import {
    getRequestHeapFrameInstruction,
    getSetComputeUnitLimitInstruction,
    getSetComputeUnitPriceInstruction,
    getSetLoadedAccountsDataSizeLimitInstruction,
} from '@solana-program/compute-budget';
import { getTransferSolInstruction } from '@solana-program/system';
import {
    address,
    appendTransactionMessageInstruction,
    blockhash,
    compileTransaction,
    createNoopSigner,
    createTransactionMessage,
    getBase64EncodedWireTransaction,
    lamports,
    pipe,
    setTransactionMessageComputeUnitLimit,
    setTransactionMessageComputeUnitPrice,
    setTransactionMessageConfig,
    setTransactionMessageFeePayer,
    setTransactionMessageLifetimeUsingBlockhash,
    type TransactionMessage,
    type TransactionMessageWithFeePayer,
} from '@solana/kit';
import { describe, expect, it } from 'vitest';

import { computeBudgetOfBase64Transaction, computeBudgetOfMessage, defaultComputeUnitLimit } from '../src/lib/budget';
import { EXAMPLE_CONFIG } from '../src/lib/v1';
import { compileInstruction, computeBudgetMessage, grpcMessage } from './fixtures';

describe('computeBudgetOfMessage', () => {
    it('should read a v1 budget straight off the config', () => {
        const budget = computeBudgetOfMessage(
            grpcMessage({
                config: {
                    computeUnitLimit: 20_000,
                    heapSize: 65_536,
                    loadedAccountsDataSizeLimit: 65_536,
                    priorityFee: '5000',
                },
                versioned: true,
            }),
        );

        expect(budget).toStrictEqual({
            computeUnitLimit: 20_000,
            heapSize: 65_536,
            loadedAccountsDataSizeLimit: 65_536,
            priorityFeeLamports: 5_000n,
        });
    });

    it('should recover a v0 budget from ComputeBudget instructions', () => {
        const budget = computeBudgetOfMessage(
            computeBudgetMessage([
                compileInstruction(getSetComputeUnitLimitInstruction({ units: 20_000 })),
                compileInstruction(getSetLoadedAccountsDataSizeLimitInstruction({ accountDataSizeLimit: 65_536 })),
                compileInstruction(getRequestHeapFrameInstruction({ bytes: 65_536 })),
            ]),
        );

        expect(budget.computeUnitLimit).toBe(20_000);
        expect(budget.loadedAccountsDataSizeLimit).toBe(65_536);
        expect(budget.heapSize).toBe(65_536);
    });

    it('should convert a v0 price in micro-lamports per unit into a total in lamports', () => {
        // 20,000 CU at 250,000 micro-lamports/CU is 5,000 lamports — the same
        // number a v1 transaction would state directly.
        const budget = computeBudgetOfMessage(
            computeBudgetMessage([
                compileInstruction(getSetComputeUnitLimitInstruction({ units: 20_000 })),
                compileInstruction(getSetComputeUnitPriceInstruction({ microLamports: 250_000n })),
            ]),
        );

        expect(budget.priorityFeeLamports).toBe(5_000n);
    });

    it('should round a fractional v0 fee up to whole lamports', () => {
        const budget = computeBudgetOfMessage(
            computeBudgetMessage([
                compileInstruction(getSetComputeUnitLimitInstruction({ units: 1 })),
                compileInstruction(getSetComputeUnitPriceInstruction({ microLamports: 1n })),
            ]),
        );

        expect(budget.priorityFeeLamports).toBe(1n);
    });

    it('should apply the implicit limit when a v0 price has no explicit limit', () => {
        const budget = computeBudgetOfMessage(
            computeBudgetMessage([
                compileInstruction(getSetComputeUnitPriceInstruction({ microLamports: 1_000_000n })),
            ]),
        );

        expect(defaultComputeUnitLimit(1)).toBe(200_000);
        expect(budget.priorityFeeLamports).toBe(200_000n);
    });

    it('should clamp the implicit limit to the maximum', () => {
        expect(defaultComputeUnitLimit(7)).toBe(1_400_000);
        expect(defaultComputeUnitLimit(100)).toBe(1_400_000);
    });

    it('should report nothing set for a transaction with no budget instructions', () => {
        expect(computeBudgetOfMessage(computeBudgetMessage([], false))).toStrictEqual({});
    });

    it('should ignore instructions from other programs', () => {
        const message = computeBudgetMessage([
            compileInstruction(getSetComputeUnitLimitInstruction({ units: 20_000 })),
        ]);
        message.accountKeys = [new Uint8Array(32).fill(7)];

        expect(computeBudgetOfMessage(message).computeUnitLimit).toBeUndefined();
    });

    it('should never fall back to instructions on a v1 message', () => {
        // A v1 message carrying a ComputeBudget instruction is malformed, but
        // the config is still the only authority — silently preferring the
        // instruction is the mix-up this accessor exists to prevent.
        const message = computeBudgetMessage([
            compileInstruction(getSetComputeUnitLimitInstruction({ units: 999_999 })),
        ]);
        message.config = { computeUnitLimit: 20_000 };

        expect(computeBudgetOfMessage(message).computeUnitLimit).toBe(20_000);
    });
});

const FEE_PAYER = address('4Nd1mBQtrMJVYVfKf2PJy9NZUZdTAsp7D4xWLs4gDB4T');
const RECIPIENT = address('9WzDXwBbmkg8ZTbNMqUxvQRAyrZzDsGYdLVL9zYtAWWM');
const LIFETIME = { blockhash: blockhash('11111111111111111111111111111111'), lastValidBlockHeight: 100n } as const;

const TRANSFER = getTransferSolInstruction({
    amount: lamports(1n),
    destination: RECIPIENT,
    source: createNoopSigner(FEE_PAYER),
});

function legacyMessage() {
    return pipe(
        createTransactionMessage({ version: 'legacy' }),
        m => setTransactionMessageFeePayer(FEE_PAYER, m),
        m => setTransactionMessageLifetimeUsingBlockhash(LIFETIME, m),
        m => appendTransactionMessageInstruction(TRANSFER, m),
    );
}

function v0Message() {
    return pipe(
        createTransactionMessage({ version: 0 }),
        m => setTransactionMessageFeePayer(FEE_PAYER, m),
        m => setTransactionMessageLifetimeUsingBlockhash(LIFETIME, m),
        m => appendTransactionMessageInstruction(TRANSFER, m),
    );
}

function v1Message() {
    return pipe(
        createTransactionMessage({ version: 1 }),
        m => setTransactionMessageFeePayer(FEE_PAYER, m),
        m => setTransactionMessageLifetimeUsingBlockhash(LIFETIME, m),
        m => appendTransactionMessageInstruction(TRANSFER, m),
    );
}

function toBase64(message: TransactionMessage & TransactionMessageWithFeePayer): string {
    return getBase64EncodedWireTransaction(compileTransaction(message));
}

describe('computeBudgetOfBase64Transaction', () => {
    it('should read a v1 budget off the config', () => {
        const encoded = toBase64(setTransactionMessageConfig(EXAMPLE_CONFIG, v1Message()));

        expect(computeBudgetOfBase64Transaction(encoded)).toStrictEqual({
            budget: { ...EXAMPLE_CONFIG },
            version: 1,
        });
    });

    it("should recover an equivalent budget from a legacy transaction's ComputeBudget instructions", () => {
        // 20,000 CU at 250,000 micro-lamports/CU is the 5,000 lamports the v1
        // config states directly.
        const encoded = toBase64(
            pipe(
                legacyMessage(),
                m => setTransactionMessageComputeUnitLimit(EXAMPLE_CONFIG.computeUnitLimit, m),
                m => setTransactionMessageComputeUnitPrice(250_000n, m),
            ),
        );

        expect(computeBudgetOfBase64Transaction(encoded)).toStrictEqual({
            budget: {
                computeUnitLimit: EXAMPLE_CONFIG.computeUnitLimit,
                priorityFeeLamports: EXAMPLE_CONFIG.priorityFeeLamports,
            },
            version: 'legacy',
        });
    });

    it("should recover an equivalent budget from a v0 transaction's ComputeBudget instructions", () => {
        const encoded = toBase64(
            pipe(
                v0Message(),
                m => setTransactionMessageComputeUnitLimit(EXAMPLE_CONFIG.computeUnitLimit, m),
                m => setTransactionMessageComputeUnitPrice(250_000n, m),
            ),
        );

        expect(computeBudgetOfBase64Transaction(encoded)).toStrictEqual({
            budget: {
                computeUnitLimit: EXAMPLE_CONFIG.computeUnitLimit,
                priorityFeeLamports: EXAMPLE_CONFIG.priorityFeeLamports,
            },
            version: 0,
        });
    });

    it('should report nothing set for a transaction that states no budget', () => {
        expect(computeBudgetOfBase64Transaction(toBase64(v0Message()))).toStrictEqual({ budget: {}, version: 0 });
    });

    it('should charge the implicit limit when a transaction prices units without limiting them', () => {
        // Two instructions: the transfer and the price itself.
        const encoded = toBase64(pipe(legacyMessage(), m => setTransactionMessageComputeUnitPrice(1_000_000n, m)));

        expect(computeBudgetOfBase64Transaction(encoded).budget.priorityFeeLamports).toBe(
            BigInt(defaultComputeUnitLimit(2)),
        );
    });
});

File: go/txv1/budget_test.go
// Offline tests for reading a compute budget across all three versions.

package txv1_test

import (
	"encoding/binary"
	"testing"

	"github.com/solana-foundation/solana-go/v2"
	computebudget "github.com/solana-foundation/solana-go/v2/programs/compute-budget"

	"github.com/solana-foundation/transaction-v1-examples/go/pb"
	"github.com/solana-foundation/transaction-v1-examples/go/txv1"
)

func computeBudgetInstruction(discriminant uint8, operand []byte) *pb.CompiledInstruction {
	return &pb.CompiledInstruction{
		ProgramIdIndex: 0,
		Data:           append([]byte{discriminant}, operand...),
	}
}

func leUint32(value uint32) []byte {
	operand := make([]byte, 4)
	binary.LittleEndian.PutUint32(operand, value)
	return operand
}

func leUint64(value uint64) []byte {
	operand := make([]byte, 8)
	binary.LittleEndian.PutUint64(operand, value)
	return operand
}

// instructionMessage is a legacy or v0 message whose only account key is the
// ComputeBudget program.
func instructionMessage(instructions []*pb.CompiledInstruction, versioned bool) *pb.Message {
	return &pb.Message{
		AccountKeys:  [][]byte{solana.ComputeBudget.Bytes()},
		Config:       nil,
		Instructions: instructions,
		Versioned:    versioned,
	}
}

func TestAV1BudgetIsReadStraightOffTheConfig(t *testing.T) {
	priorityFee, computeUnitLimit, heapSize := uint64(5_000), uint32(20_000), uint32(65_536)
	loadedAccounts := uint32(65_536)
	message := &pb.Message{
		Config: &pb.TransactionConfig{
			ComputeUnitLimit:            &computeUnitLimit,
			HeapSize:                    &heapSize,
			LoadedAccountsDataSizeLimit: &loadedAccounts,
			PriorityFee:                 &priorityFee,
		},
		Versioned: true,
	}

	budget := txv1.BudgetOfMessage(message)

	assertUint32(t, "computeUnitLimit", budget.ComputeUnitLimit, 20_000)
	assertUint32(t, "heapSize", budget.HeapSize, 65_536)
	assertUint32(t, "loadedAccountsDataSizeLimit", budget.LoadedAccountsDataSizeLimit, 65_536)
	assertUint64(t, "priorityFee", budget.PriorityFee, 5_000)
}

func TestAV0BudgetIsRecoveredFromComputeBudgetInstructions(t *testing.T) {
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitLimit, leUint32(20_000)),
		computeBudgetInstruction(computebudget.Instruction_SetLoadedAccountsDataSizeLimit, leUint32(65_536)),
		computeBudgetInstruction(computebudget.Instruction_RequestHeapFrame, leUint32(65_536)),
	}, true)

	budget := txv1.BudgetOfMessage(message)

	assertUint32(t, "computeUnitLimit", budget.ComputeUnitLimit, 20_000)
	assertUint32(t, "loadedAccountsDataSizeLimit", budget.LoadedAccountsDataSizeLimit, 65_536)
	assertUint32(t, "heapSize", budget.HeapSize, 65_536)
}

func TestAV0PriorityFeeIsConvertedFromMicroLamportsPerUnitToATotal(t *testing.T) {
	// 20,000 CU at 250,000 micro-lamports/CU = 5,000,000,000 micro-lamports,
	// which is the 5,000 lamports a v1 transaction would state directly.
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitLimit, leUint32(20_000)),
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitPrice, leUint64(250_000)),
	}, true)

	assertUint64(t, "priorityFee", txv1.BudgetOfMessage(message).PriorityFee, 5_000)
}

func TestAV0PriorityFeeRoundsUpToWholeLamports(t *testing.T) {
	// 1 CU at 1 micro-lamport is a millionth of a lamport, which the runtime
	// charges as one lamport rather than as zero.
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitLimit, leUint32(1)),
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitPrice, leUint64(1)),
	}, true)

	assertUint64(t, "priorityFee", txv1.BudgetOfMessage(message).PriorityFee, 1)
}

func TestAV0PriceWithoutAnExplicitLimitUsesTheImplicitDefault(t *testing.T) {
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitPrice, leUint64(1_000_000)),
	}, true)

	// One instruction, no SetComputeUnitLimit, so the limit is the 200k default
	// and a price of one lamport per CU totals 200,000 lamports.
	if got := txv1.DefaultComputeUnitLimit(1); got != 200_000 {
		t.Errorf("default limit for one instruction: got %d, want 200000", got)
	}
	assertUint64(t, "priorityFee", txv1.BudgetOfMessage(message).PriorityFee, 200_000)
}

func TestTheImplicitLimitIsClampedToTheMaximum(t *testing.T) {
	for _, instructionCount := range []int{7, 100} {
		if got := txv1.DefaultComputeUnitLimit(instructionCount); got != 1_400_000 {
			t.Errorf("%d instructions: got %d, want 1400000", instructionCount, got)
		}
	}
}

func TestATransactionWithNoBudgetInstructionsReportsNothingSet(t *testing.T) {
	budget := txv1.BudgetOfMessage(instructionMessage(nil, false))

	if !txv1.ConfigsEqual(budget, solana.TransactionConfig{}) {
		t.Errorf("expected an empty budget, got %+v", budget)
	}
}

func TestInstructionsFromOtherProgramsAreIgnored(t *testing.T) {
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitLimit, leUint32(20_000)),
	}, true)
	// Point the instruction at a different program without changing its data.
	message.AccountKeys[0] = make([]byte, solana.PublicKeyLength)

	if limit := txv1.BudgetOfMessage(message).ComputeUnitLimit; limit != nil {
		t.Errorf("only the ComputeBudget program's instructions carry a budget, got %d", *limit)
	}
}

func TestAV1MessageNeverFallsBackToScanningInstructions(t *testing.T) {
	// A v1 message carrying a ComputeBudget instruction is malformed, but the
	// config is still the only authority: silently preferring the instruction is
	// exactly the mix-up this accessor exists to prevent.
	message := instructionMessage([]*pb.CompiledInstruction{
		computeBudgetInstruction(computebudget.Instruction_SetComputeUnitLimit, leUint32(999_999)),
	}, true)
	computeUnitLimit := uint32(20_000)
	message.Config = &pb.TransactionConfig{ComputeUnitLimit: &computeUnitLimit}

	assertUint32(t, "computeUnitLimit", txv1.BudgetOfMessage(message).ComputeUnitLimit, 20_000)
}
